### PR TITLE
User end points

### PR DESCRIPTION
**Pull request by azharmalik3**
@mosic

### CODE REVIEW

File: lib/evercam_media/util.ex [hunk 1]
```diff
@@ -116,4 +117,33 @@ defmodule EvercamMedia.Util do
...
     |> Calendar.DateTime.Format.unix
   end
+
+  def user_at_intercom(user) do
+    intercomUser = HTTPotion.get "#{System.get_env["INTERCOM_URL"]}?user_id=#{user.username}",
+                    [
+                      basic_auth: {"#{System.get_env["INTERCOM_ID"]}", "#{System.get_env["INTERCOM_KEY"]}"},
+                      headers: ["Accept": "Accept:application/json"]
+                    ]
+    case HTTPotion.Response.success?(intercomUser) do
```
Comment (mosic) [inline lib/evercam_media/util.ex]: Always use snake_case for variable names: https://github.com/niftyn8/elixir_style_guide#naming

File: lib/evercam_media/util.ex [hunk 2]
```diff
@@ -116,4 +117,33 @@ defmodule EvercamMedia.Util do
     |> Calendar.DateTime.from_erl!("Etc/UTC")
     |> Calendar.DateTime.Format.unix
   end
+
+  def user_at_intercom(user) do
+    intercomUser = HTTPotion.get "#{System.get_env["INTERCOM_URL"]}?user_id=#{user.username}",
```
Comment (mosic) [inline lib/evercam_media/util.ex]: This block looks really strange. Use parentheses around function arguments, and don't line up multiple lines in this way.

File: lib/evercam_media/util.ex [hunk 3]
```diff
@@ -116,4 +117,33 @@ defmodule EvercamMedia.Util do
...
+                    [
+                      basic_auth: {"#{System.get_env["INTERCOM_ID"]}", "#{System.get_env["INTERCOM_KEY"]}"},
```
Comment (mosic) [inline lib/evercam_media/util.ex]: There is no need to put strings into string interpolation.

This:
```
{"#{System.get_env["INTERCOM_ID"]}", "#{System.get_env["INTERCOM_KEY"]}"}
```
can be written as

```
{System.get_env["INTERCOM_ID"], System.get_env["INTERCOM_KEY"]}
```

File: lib/evercam_media/util.ex [hunk 4]
```diff
@@ -116,4 +117,33 @@ defmodule EvercamMedia.Util do
...
+      false -> {:error, intercomUser}
+    end
+  end
+
+  def create_intercom_user(intercom_user) do
+    json = case Poison.encode intercom_user do
+      {:ok, json} -> json
+      _ -> nil
+    end
+    intercomUser = HTTPotion.post "#{System.get_env["INTERCOM_URL"]}",
```
Comment (mosic) [inline lib/evercam_media/util.ex]: It would be nicer to extract this as a module variable as it's being used in multiple places. Like this:

``` elixir
@intercom_url System.get_env["INTERCOM_URL"]
@intercom_auth {System.get_env["INTERCOM_ID"], System.get_env["INTERCOM_KEY"]}
```

File: lib/evercam_media/util.ex [hunk 5]
```diff
@@ -116,4 +117,33 @@ defmodule EvercamMedia.Util do
...
+                      headers: ["Accept": "Accept:application/json"]
+                    ]
+    case HTTPotion.Response.success?(intercomUser) do
+      true -> {:ok, intercomUser}
+      false -> {:error, intercomUser}
+    end
+  end
+
+  def create_intercom_user(intercom_user) do
+    json = case Poison.encode intercom_user do
```
Comment (mosic) [inline lib/evercam_media/util.ex]: Put the `case` on a new line, like we do for asignment with multiline pipelines: https://github.com/niftyn8/elixir_style_guide

File: lib/evercam_media/util.ex [hunk 6]
```diff
@@ -116,4 +117,33 @@ defmodule EvercamMedia.Util do
...
+      {:ok, json} -> json
+      _ -> nil
+    end
+    intercomUser = HTTPotion.post "#{System.get_env["INTERCOM_URL"]}",
+                    [
+                      body: json,
+                      basic_auth: {"#{System.get_env["INTERCOM_ID"]}", "#{System.get_env["INTERCOM_KEY"]}"},
+                      headers: ["Accept": "Accept:application/json", "Content-Type": "application/json"]
+                    ]
+    case HTTPotion.Response.success?(intercomUser) do
```
Comment (mosic) [inline lib/evercam_media/util.ex]: This seems unnecessary, just return the result of the http request.

File: web/controllers/user_controller.ex [hunk 7]
```diff
@@ -24,4 +31,200 @@ defmodule EvercamMedia.UserController do
...
+      changeset = User.changeset(%User{}, params)
+      case Repo.insert(changeset) do
+        {:ok, user} ->
+          {:ok, exp_date } = Calendar.DateTime.now!("UTC") |> Calendar.DateTime.advance(3600)
+          {:ok, expiry_date } = Calendar.DateTime.to_erl(exp_date) |> Ecto.DateTime.cast
+          token = Ecto.build_assoc(user, :access_tokens, is_revoked: false,
+            request: UUID.uuid4(:hex) |> String.slice(0..15), expires_at: expiry_date )
+
+          case Repo.insert(token) do
+            { :ok, token } -> { :success, user, token }
```
Comment (mosic) [inline web/controllers/user_controller.ex]: Don't put spaces inside braces/any type of brackets: https://github.com/niftyn8/elixir_style_guide#source-code-layout

File: web/controllers/user_controller.ex [hunk 8]
```diff
@@ -24,4 +31,200 @@ defmodule EvercamMedia.UserController do
...
+              "last_request_at": user.created_at |> Util.ecto_datetime_to_unix,
+              "last_seen_ip": requester_ip,
+              "signed_up_at": user.created_at |> Util.ecto_datetime_to_unix
+            }
+            case Util.create_intercom_user(intercom_user) do
+              true -> conn |> render(UserView, "show.json", %{user: user |> Repo.preload(:country, force: true)})
+              error -> conn |> put_status(404) |> render(ErrorView, "error.json", %{message: error})
+            end
+          end
+        {:error, changeset} -> conn |> put_status(404) |> render(ErrorView, "error.json", %{message: changeset})
```
Comment (mosic) [inline web/controllers/user_controller.ex]: Put `conn...` on a new line, like it was done with code in `{:ok, user}` case.

File: web/models/country.ex [hunk 9]
```diff
@@ -11,6 +13,21 @@ defmodule Country do
...
+    Country
+    |> where(iso3166_a2: ^country_id)
+    |> Repo.one
+  end
+
+  def get_by_code(country_id) when country_id in [nil, ""] do
+    {:error, nil}
+  end
+  def get_by_code(country_id) do
+    if country = Repo.get_by(Country, iso3166_a2: country_id) do
```
Comment (mosic) [inline web/models/country.ex]: And what if `country_id` doesn't fit this clause? Cover all cases.

File: web/models/user.ex [hunk 10]
```diff
@@ -23,8 +23,9 @@ defmodule User do
     field :token_expires_at, Ecto.DateTime
     field :stripe_customer_id, :string
     field :confirmed_at, Ecto.DateTime
-    field :updated_at, Ecto.DateTime, default: Ecto.DateTime.utc
```
Comment (mosic) [inline web/models/user.ex]: Why was this change necessary?

File: web/validations/user.ex [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        _ -> {:invalid, "The email is already registered."}
+      end
+    else
+      invalid(:email)
+    end
+  end
+
+  defp validate(:username, value) do
+    if Regex.match?(@username_pattern, value) && String.length(value) > 3 do
+      from(u in User)
```
Comment (mosic) [inline web/validations/user.ex]: `from(u in User)` is unnecessary, just use `User`.

File: web/controllers/user_controller.ex [hunk 12]
```diff
@@ -24,4 +31,200 @@ defmodule EvercamMedia.UserController do
...
+      conn
+      |> render(UserView, "credentials.json", %{user: user})
+    end
+  end
+
+  def create(conn, params) do
+    with :ok <- ensure_country(params["country_id"], conn),
+         :ok <- validate_params(params) |> ensure_params(conn)
+    do
+      [user_agent | rest] = conn |> get_req_header("user-agent")
```
Comment (mosic) [inline web/controllers/user_controller.ex]: Avoid using single pipe without a good reason, replace this with `get_req_header(conn, "user-agent")`.

Comment (mosic) [inline web/controllers/user_controller.ex]: Don't put spaces around `|`.

File: web/controllers/user_controller.ex [hunk 13]
```diff
@@ -2,6 +2,13 @@ defmodule EvercamMedia.UserController do
   use EvercamMedia.Web, :controller
   alias EvercamMedia.UserView
   alias EvercamMedia.ErrorView
+  alias EvercamMedia.Repo
+  import EvercamMedia.Validation.User
+  alias EvercamMedia.Util
```
Comment (mosic) [inline web/controllers/user_controller.ex]: Group `alias`, `use` etc. lines together, don't mix them up.

File: lib/evercam_media/util.ex [hunk 14]
```diff
@@ -2,6 +2,7 @@ defmodule EvercamMedia.Util do
   use Calendar
   require Logger
   import String, only: [to_integer: 1]
+  require IEx
```
Comment (mosic) [inline lib/evercam_media/util.ex]: Remove this.

File: web/controllers/user_controller.ex [hunk 15]
```diff
@@ -24,4 +31,200 @@ defmodule EvercamMedia.UserController do
...
+  defp firstname(firstname, _user),  do: firstname
+
+  defp lastname(lastname, user) when lastname in [nil, ""], do: user.lastname
+  defp lastname(lastname, _user), do: lastname
+
+  defp email(email, user) when email in [nil, ""], do: user.email
+  defp email(email, _user), do: email
+
+  defp country(country_id, user) when country_id in [nil, ""] do
+    IEx.pry
```
Comment (mosic) [inline web/controllers/user_controller.ex]: You have left the debugger on.

File: web/validations/user.ex [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+
+  defp validate(:email, value) do
+    if Regex.match?(@email_pattern, value) do
+      from(u in User)
+      |> select([u], count(u.id))
+      |> where([u], ilike(u.email, ^value))
+      |> Repo.one
+      |> case do
+        0 -> :ok
+        _ -> {:invalid, "The email is already registered."}
```
Comment (mosic) [inline web/validations/user.ex]: It would be better to use `Ecto.Changeset` validations for validating this: https://hexdocs.pm/ecto/2.0.0-rc.5/Ecto.Changeset.html

**Comment by mosic** [comment]
@ijunaidfarooq There were repeated issues for some things (e.g. `case` on the same line as `=`), I didn't leave a comment for every instance of them, just the first one I spotted.

File: web/models/country.ex (hunk 9 above)
Comment (ijunaidfarooq) [inline web/models/country.ex]: this thing already checked in `ensure_country` its only coming to this point when `country_id` belongs to some country

File: web/models/user.ex (hunk 10 above)
Comment (ijunaidfarooq) [inline web/models/user.ex]: updated as wasn't working so I added the concept of `timestamps` and it was handling updated at

Comment (mosic) [inline web/models/user.ex]: That sounds weird, it should be the same thing. Try it again and let me know why it doesn't work. If it does work, leave it as it was.

File: web/models/country.ex (hunk 9 above)
Comment (mosic) [inline web/models/country.ex]: Then why use `if` here?

File: web/models/user.ex (hunk 10 above)
Comment (ijunaidfarooq) [inline web/models/user.ex]: Okay will tell you about this!

File: web/models/country.ex (hunk 9 above)
Comment (ijunaidfarooq) [inline web/models/country.ex]: I wanted that pattern 

```
{:ok, country}
```
is it possible to do something like
`{:ok, country} = Repo.get_by(Country, iso3166_a2: country_id)`

Comment (mosic) [inline web/models/country.ex]: Then you can use:

``` elixir
country = Repo.get_by(Country, iso3166_a2: country_id)
{:ok, country}
```

But I'd instead refactor this to cover all cases, including invalid country codes:

Before:
``` elixir
def get_by_code(country_id) when country_id in [nil, ""] do
  {:error, nil}
end
def get_by_code(country_id) do
  if country = Repo.get_by(Country, iso3166_a2: country_id) do
    {:ok, country}
  end
end
```

After:
``` elixir
def get_by_code(country_id) when country_id in [nil, ""], do: {:error, nil}
def get_by_code(country_id) do
  case Repo.get_by(Country, iso3166_a2: country_id) do
    nil -> {:error, nil}
    country -> {:ok, country}
  end
end
```

File: web/models/user.ex (hunk 10 above)
Comment (ijunaidfarooq) [inline web/models/user.ex]: I have removed this line

`# timestamps(inserted_at: false, type: Ecto.DateTime, default: Ecto.DateTime.utc)`
and added the previous one back

`field :updated_at, Ecto.DateTime, default: Ecto.DateTime.utc`

results of updating were

with timestamps 

```
begin []
[debug] QUERY OK db=26.3ms
UPDATE "users" SET "email" = $1, "updated_at" = $2 WHERE "id" = $3 ["junaid2@evercam.io", {{2016, 6, 10}, {6, 34, 57, 0}}, 11172]
[debug] QUERY OK db=38.4ms
commit []
[debug] QUERY OK db=2.6ms decode=0.1ms queue=0.4ms
SELECT c0."id", c0."iso3166_a2", c0."name", c0."id" FROM "countries" AS c0 WHERE (c0."id" = ANY($1)) [[175]]
```

without timestamps and the previous line
```
iex(1)> [debug] QUERY OK db=0.5ms queue=0.1ms
begin []
[debug] QUERY OK db=2.4ms
UPDATE "users" SET "email" = $1 WHERE "id" = $2 ["c@evercam.io", 11172]
[debug] QUERY OK db=50.0ms
commit []
[debug] QUERY OK db=1.1ms queue=0.2ms
SELECT c0."id", c0."iso3166_a2", c0."name", c0."id" FROM "countries" AS c0 WHERE (c0."id" = ANY($1)) [[175]]
```

same as in Database.. more references from https://groups.google.com/forum/#!topic/phoenix-talk/4lRdIg6NKdo

Comment (mosic) [inline web/models/user.ex]: So it's just that the `updated_at` needs to be explicitly given to the changeset.

Ok, we can use `timestamps` to make the code a bit simpler. But it would be better to cover both `updated_at` and `created_at`, like this:

``` elixir
timestamps(inserted_at: :created_at, type: Ecto.DateTime, default: Ecto.DateTime.utc)
```

Comment (mosic) [inline web/models/user.ex]: Update: because this needed to be changed in all models (and related code), I went ahead and did it in https://github.com/evercam/evercam-media/commit/a725a5fe55c841ccaf4fc7d33c74bf38b5d724d7.

You can either remove your change or use the line from the comment above, both should merge cleanly.

**Comment by mosic** [comment]
@ijunaidfarooq My comment about validation ([files#r66448004](https://github.com/evercam/evercam-media/pull/250/files#r66448004)) still stands, use Ecto to validate user data: https://hexdocs.pm/ecto/2.0.0-rc.5/Ecto.Changeset.html

**Comment by ijunaidfarooq** [comment]
Only Validation is working through `changeset`. 
as per docs https://hexdocs.pm/ecto/2.0.0-rc.5/Ecto.Changeset.html#unique_constraint/3

In order to use the uniqueness constraint the first step is to define the unique index in a migration.. in our case it's going to be the database? 

as now uniqueness isn't working with this

```
    |> unique_constraint(:email, [name: "ux_users_email"])
    |> unique_constraint(:username, [name: "ux_users_username"])
```

File: web/controllers/user_controller.ex [hunk 17]
```diff
@@ -8,6 +8,7 @@ defmodule EvercamMedia.UserController do
   import EvercamMedia.Validation.User
   import Ecto.Changeset
   import Ecto.Query
+  require IEx
```
Comment (mosic) [inline web/controllers/user_controller.ex]: Remove debugging code.

File: web/controllers/user_controller.ex [hunk 18]
```diff
@@ -98,6 +99,16 @@ defmodule EvercamMedia.UserController do
               |> Base.encode16
               |> String.downcase
 
+            evercam_user = User.by_username("evercam")
+            IEx.pry
```
Comment (mosic) [inline web/controllers/user_controller.ex]: Remove debugging code.

File: web/controllers/user_controller.ex [hunk 19]
```diff
@@ -98,6 +99,16 @@ defmodule EvercamMedia.UserController do
               |> Base.encode16
               |> String.downcase
 
+            evercam_user = User.by_username("evercam")
```
Comment (mosic) [inline web/controllers/user_controller.ex]: It would be nicer if adding the default camera share was in a separate function.

**Comment by mosic** [comment]
@ijunaidfarooq Good question. I'll check and get back to you tomorrow.

**Comment by mosic** [comment]
@ijunaidfarooq You were right, the uniqueness validation wasn't working with the schema.

I added unique indexes, everything works on the master branch now. Example:

``` elixir
iex(evercam@shade)1> User.changeset(%User{}, %{email: "jonsnow@example.com", firstname: "n", lastname: "Snow", password: "nothing", username: "jonsnow", country_id: 1}) |> EvercamMedia.Repo.insert
[debug] QUERY OK db=0.2ms
begin []
[debug] QUERY OK db=1.2ms
INSERT INTO "users" ("country_id","email","firstname","lastname","password","username","created_at","updated_at") VALUES ($1,$2,$3,$4,$5,$6,$7,$8) RETURNING "id" [1, "jonsnow@example.com", "n", "Snow", "nothing", "jonsnow", {{2016, 6, 16}, {23, 28, 30, 0}}, {{2016, 6, 16}, {23, 28, 30, 0}}]
[debug] QUERY OK db=2.2ms
commit []
{:ok,
 %User{__meta__: #Ecto.Schema.Metadata<:loaded, "users">,
  access_tokens: #Ecto.Association.NotLoaded<association :access_tokens is not loaded>,
  api_id: nil, api_key: nil, billing_id: nil,
  camera_shares: #Ecto.Association.NotLoaded<association :camera_shares is not loaded>,
  cameras: #Ecto.Association.NotLoaded<association :cameras is not loaded>,
  confirmed_at: nil,
  country: #Ecto.Association.NotLoaded<association :country is not loaded>,
  country_id: 1, created_at: #Ecto.DateTime<2016-06-16 23:28:30>,
  email: "jonsnow@example.com", firstname: "n", id: 16164, lastname: "Snow",
  password: "nothing", stripe_customer_id: nil, token_expires_at: nil,
  updated_at: #Ecto.DateTime<2016-06-16 23:28:30>, username: "jonsnow"}}

iex(evercam@shade)2> User.changeset(%User{}, %{email: "jonsnow@example.com", firstname: "n", lastname: "Snow", password: "nothing", username: "jonsnow", country_id: 1}) |> EvercamMedia.Repo.insert
[debug] QUERY OK db=0.1ms
begin []
[debug] QUERY ERROR db=1.4ms
INSERT INTO "users" ("country_id","email","firstname","lastname","password","username","created_at","updated_at") VALUES ($1,$2,$3,$4,$5,$6,$7,$8) RETURNING "id" [1, "jonsnow@example.com", "n", "Snow", "nothing", "jonsnow", {{2016, 6, 16}, {23, 28, 35, 0}}, {{2016, 6, 16}, {23, 28, 35, 0}}]
[debug] QUERY OK db=0.1ms
rollback []
{:error,
 #Ecto.Changeset<action: :insert,
  changes: %{country_id: 1, email: "jonsnow@example.com", firstname: "n",
    lastname: "Snow", password: "nothing", username: "jonsnow"},
  errors: [email: {"has already been taken", []}], data: #User<>,
  valid?: false>}
```

**Comment by mosic** [comment]
@ijunaidfarooq Btw you can setup the local db with updated schema with:

```
mix do ecto.drop, ecto.create, ecto.load
```

**Comment by ijunaidfarooq** [comment]
@mosic nice example data.. :)

will this `ecto.load` will load up all data same as we get from `./scripts/live_clone`?

File: lib/evercam_media/util.ex [hunk 20]
```diff
@@ -2,6 +2,7 @@ defmodule EvercamMedia.Util do
   use Calendar
   require Logger
   import String, only: [to_integer: 1]
+  import Ecto.Changeset
```
Comment (mosic) [inline lib/evercam_media/util.ex]: It would be better if this was `import Ecto.Changeset, only: [traverse_errors: 2]`, so it's obvious where the `traverse_errors` function came from.

In general, always prefer either explicit imports or using full name (`Ecto.Changeset.traverse_errors(...)`).

File: web/models/user.ex [hunk 21]
```diff
@@ -82,6 +82,7 @@ defmodule User do
     |> cast(params, @required_fields, @optional_fields)
     |> unique_constraint(:username, [name: :user_username_unique_index])
     |> unique_constraint(:email, [name: :user_email_unique_index])
-    |> validate_format(:email, ~r/^.+@.+\..+$/)
+    |> validate_format(:username, ~r/^[a-z]+[\w-]+$/)
+    |> validate_format(:email, ~r/\A([\w+\-]\.?)+@[a-z\d\-]+(\.[a-z]+)*\.[a-z]+\z/)
```
Comment (mosic) [inline web/models/user.ex]: Revert to the simpler version of regex for email validation, it's enough to validate that it's in the format of `{string}@{string}.{string}`.

File: web/validations/user.ex [hunk 22]
```diff
@@ -0,0 +1,16 @@
+defmodule EvercamMedia.Validation.User do
+  def validate_params(params) do
```
Comment (mosic) [inline web/validations/user.ex]: Checking for empty parameters is now handled by the changeset, you can remove this module.

File: lib/evercam_media/util.ex [hunk 23]
```diff
@@ -127,4 +129,32 @@ defmodule EvercamMedia.Util do
     |> Calendar.DateTime.from_erl!("Etc/UTC")
     |> Calendar.DateTime.Format.unix
   end
+
+  def user_at_intercom(user) do
```
Comment (mosic) [inline lib/evercam_media/util.ex]: Move these functions to a separate module, `EvercamMedia.Intercom`, in `lib/evercam_media/intercom.ex`.

Also, `Intercom.get_user(username)` and  `Intercom.create_user(user)` names are more in line with the rest of the code.

File: web/controllers/user_controller.ex [hunk 24]
```diff
@@ -24,4 +29,252 @@ defmodule EvercamMedia.UserController do
...
+              |> Base.encode16
+              |> String.downcase
+
+            default_camera_share(user)
+            EvercamMedia.UserMailer.confirm(user, code)
+          end
+          if user |> Util.user_at_intercom |> intercom_user? do
+            conn |> put_status(400) |> render(ErrorView, "error.json", %{message: "User already present at intercom."})
+          else
+            intercom_user = %{
```
Comment (mosic) [inline web/controllers/user_controller.ex]: Let the function which creates the Intercom user handle this, it's messy to do construct parameters here.

File: web/controllers/user_controller.ex [hunk 25]
```diff
@@ -24,4 +29,252 @@ defmodule EvercamMedia.UserController do
...
+            {:error, changeset} -> {:invalid_token, changeset}
+          end
+          if !has_share_request_key?(share_request_key) do
+            created_at =
+              user.created_at
+              |> Ecto.DateTime.to_erl
+              |> Strftime.strftime!("%Y-%m-%d %T UTC")
+
+            code =
+              :sha
```
Comment (mosic) [inline web/controllers/user_controller.ex]: It might be clearer to start the pipeline with `:crypto.hash(:sha, user.username <> created_at)`.

File: web/controllers/user_controller.ex [hunk 26]
```diff
@@ -24,4 +29,252 @@ defmodule EvercamMedia.UserController do
...
+              "user_id": user.username,
+              "name": user.firstname <> " " <> user.lastname,
+              "last_seen_user_agent": user_agent,
+              "last_request_at": user.created_at |> Util.ecto_datetime_to_unix,
+              "last_seen_ip": requester_ip,
+              "signed_up_at": user.created_at |> Util.ecto_datetime_to_unix
+            }
+            case Util.create_intercom_user(intercom_user) do
+              true -> conn |> render(UserView, "show.json", %{user: user |> Repo.preload(:country, force: true)})
+              error -> conn |> put_status(404) |> render(ErrorView, "error.json", %{message: error})
```
Comment (mosic) [inline web/controllers/user_controller.ex]: This doesn't seem right. There is no reason to return an error to the user if we failed to create a record in Intercom. We can do it in the background and log & retry the request.

File: web/controllers/user_controller.ex [hunk 27]
```diff
@@ -24,4 +29,252 @@ defmodule EvercamMedia.UserController do
...
+      |> String.replace_trailing(".json", "")
+      |> String.downcase
+      |> User.by_username
+
+    with :ok <- ensure_user_exists(user, username, conn),
+         :ok <- ensure_can_view(current_user, user, conn)
+    do
+      spawn(fn -> delete_user_worker(user) end)
+      conn
+      |> put_status(200) |> json(%{message: "User has been deleted!"})
```
Comment (mosic) [inline web/controllers/user_controller.ex]: Continue the pipeline below.

**Comment by mosic** [comment]
@ijunaidfarooq I thought you'd like it :)

No, that script just sets up the database schema, it doesn't clone any data. We can add a separate task for that.

**Comment by mosic** [comment]
@ijunaidfarooq The PR is mostly fine, but it's too big to test and merge like this. Please split it up into separate endpoints. You can extract just one (e.g. `DELETE`), and then base code off master.

File: web/controllers/user_controller.ex [hunk 28]
```diff
@@ -24,4 +29,252 @@ defmodule EvercamMedia.UserController do
...
+  defp intercom_user?({:error, _}), do: false
+
+  defp delete_user_worker(user) do
+    cameras =
+      Camera
+      |> where([cam], cam.owner_id == ^user.id)
+      |> Repo.all
+
+    delete_user_shares(user)
+    if user_owned_any_camera?(cameras) do
```
Comment (mosic) [inline web/controllers/user_controller.ex]: This is confusing. Why not delete all user's cameras in the previous step?

File: web/controllers/user_controller.ex [hunk 29]
```diff
@@ -24,4 +29,252 @@ defmodule EvercamMedia.UserController do
...
+  defp user_owned_any_camera?([]), do: false
+  defp user_owned_any_camera?(_), do: true
+
+  defp delete_user_shares(user) do
+    CameraShare
+    |> where(user_id: ^user.id)
+    |> Repo.delete_all
+  end
+
+  defp default_camera_share(user) do
```
Comment (mosic) [inline web/controllers/user_controller.ex]: This would be slightly nicer as `share_default_camera` instead of `default_camera_share`.

File: web/models/user.ex (hunk 21 above)
Comment (ijunaidfarooq) [inline web/models/user.ex]: old regex validates this as a valid email `junai d@evercam.io`

```
  Parameters: %{"email" => "junai d@evercam.io", "firstname" => "jhjhj", "lastname" => "hello", "password" => "[FILTERED]", "username" => "hjhjh"}
  Pipelines: [:api]
[debug] QUERY OK db=0.3ms queue=0.1ms
begin []
[debug] QUERY OK db=75.2ms
INSERT INTO "users" ("api_id","api_key","email","firstname","lastname","password","username","created_at","updated_at") VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9) RETURNING "id" ["3da7df62", "55d0b1a8eb494f0aa28609319eef3b63", "junai d@evercam.io", "jhjhj", "hello", "$2b$12$0bKPbRgImR0cEifd1XXtkOXj4i66pZ0lut0v2T/a2TcJF1FFBPMtq", "hjhjh", {{2016, 6, 20}, {6, 4, 52, 0}}, {{2016, 6, 20}, {6, 4, 52, 0}}]
[debug] QUERY OK db=48.1ms
commit []
```

File: web/controllers/user_controller.ex (hunk 26 above)
Comment (ijunaidfarooq) [inline web/controllers/user_controller.ex]: Should I use `spawn` for the same this thing? or monitor it with [Supervisor](http://elixir-lang.org/getting-started/mix-otp/supervisor-and-application.html#our-first-supervisor) ?

Comment (mosic) [inline web/controllers/user_controller.ex]: `spawn` would be acceptable, we can improve it if necessary.

File: web/models/user.ex (hunk 21 above)
Comment (mosic) [inline web/models/user.ex]: Yes, but the new regex rejects email addresses with special characters. I made a new regex with a couple of tests now, use that instead: https://github.com/evercam/evercam-media/blob/master/test/models/user_test.exs

(Meaning, base the PR update off master)
